File: launcher.zig
// zig fmt: off

const std = @import("std");
const builtin = @import("builtin");

const isWindows = builtin.os.tag == .windows;

const SEP: []const u8 = if (isWindows) "\\" else "/";

const INSTALLER = if (isWindows) "clisw-installer.exe" else "clisw-installer";
const LAUNCHER = if (isWindows) "clisw-launcher.exe" else "clisw-launcher";
const GAME = if (isWindows) "clisw.exe" else "clisw";

const INSTALLER_DIR = "..";
const DATA_DIR = "data";

// Array of data directories
const DATA_DIRS = &[_][]const u8{
	"misc",
	"maps",
	"saves",
	"story",
	"story" ++ SEP ++ "best_showtime",
	"story" ++ SEP ++ "control_zone",
	"story" ++ SEP ++ "r_square",
	"story" ++ SEP ++ "tower_of_greed"
};

const BAD_EXIT: i32 = 0xff;

var outBuffer: [1024]u8 = undefined;
var errBuffer: [1024]u8 = undefined;
var w: std.fs.File.Writer = switch (builtin.os.tag) {
	.windows => undefined,
	else => std.fs.File.stdout().writer(&outBuffer),
};
var e: std.fs.File.Writer = switch (builtin.os.tag) {
	.windows => undefined,
	else => std.fs.File.stderr().writer(&errBuffer),
};
const stdout = &w.interface;
const stderr = &e.interface;


fn ssleep(ms: u64) void {
	std.Thread.sleep(ms * std.time.ns_per_ms);
}

fn trimNewline(s: []u8) []u8 {
	var end = s.len;
	while (end > 0) : (end -= 1) {
		const c = s[end - 1];
		if (c == '\n' or c == '\r') continue;
		break;
	}
	return s[0..end];
}

fn checkLatest(allocator: std.mem.Allocator) !bool {
	const url = "https://raw.githubusercontent.com/jgarcdev/TextAdventure-CLISoulWorker/main/version";

	var client = std.http.Client{ .allocator = allocator };
	defer client.deinit();

	const uri = try std.Uri.parse(url);

	var resBuffer: [64]u8 = undefined;
	var responseWriter = std.io.Writer.fixed(&resBuffer);


	const res = client.fetch(.{
		.location = .{ .uri = uri },
		.redirect_behavior = .init(1),
		.response_writer = &responseWriter,
		.method = .GET,
		.headers = .{
			.user_agent = .{ .override = "VersionChecker/1.0" },
		}
	}) catch |err| {
		// std.debug.print("HTTP request failed: {}\n", .{err});
		try stderr.print("HTTP request failed: {}\n", .{err});
		try stderr.flush();
		return true;
	};

	if (res.status != .ok) {
		// std.debug.print("HTTP request returned non-OK status: {}\n", .{res.status});
		try stderr.print("HTTP request returned non-OK status: {}\n", .{res.status});
		try stderr.flush();
		return true;
	}
	const body = trimNewline(resBuffer[0..responseWriter.end]);

	// std.debug.print("Response: {s}\n", .{body});


	const localFile = std.fs.cwd().openFile("version", .{ .mode = .read_only }) catch |err| {
		// std.debug.print("Could not open local version file: {}\n", .{err});
		try stderr.print("Could not open local version file: {}\n", .{err});
		try stderr.flush();
		return true;
	};
	defer localFile.close();


	var localBuf: [32]u8 = undefined;
	const readBytes = localFile.read(&localBuf) catch |err| {
		std.debug.print("Could not read local version file: {}\n", .{err});
		try stderr.print("Could not read local version file: {}\n", .{err});
		try stderr.flush();
		return true;
	};

	const local = trimNewline(localBuf[0..readBytes]);

	try stdout.print("Checking version...\nCurrent version is {s}\nLatest version is {s}\n", .{local, body});
	try stdout.flush();

	return std.mem.eql(u8, body, local);
}

const Opt = enum { FIX, UPDATE };

fn runInstaller(allocator: std.mem.Allocator, opt: Opt) !noreturn {
	// std.debug.print("Changing to installer directory...\n", .{});

	var installerDir = std.fs.cwd().openDir(INSTALLER_DIR, .{}) catch |err| {
		// std.debug.print("Could not open installer directory: {}\n", .{err});
		try stderr.print("Could not open installer directory: {}\n", .{err});
		try stderr.flush();
		std.process.exit(BAD_EXIT);
	};
	defer installerDir.close();
	installerDir.setAsCwd() catch |err| {
		// std.debug.print("Could not change to installer directory: {}\n", .{err});
		try stderr.print("Could not change to installer directory: {}\n", .{err});
		try stderr.flush();
		std.process.exit(BAD_EXIT);
	};

	ssleep(1000);

	const args = switch (opt) {
		Opt.FIX => &[_][]const u8{INSTALLER, "-f"},
		Opt.UPDATE => &[_][]const u8{INSTALLER, "-u"},
	};

	var proc = std.process.Child.init(args, allocator);
	try proc.spawn();

	const exitCode = proc.wait() catch |err| {
		try stdout.print("Could not execute installer: {}\n", .{err});
		try stdout.flush();
		std.process.exit(BAD_EXIT);
	};
	if (exitCode.Exited != 0) {
		try stderr.print("Installer exited with code {}\n", .{exitCode});
		try stderr.flush();
		std.process.exit(BAD_EXIT);
	}
	std.process.exit(0);
}

fn verifyDataDir(allocator: std.mem.Allocator, cwd: std.fs.Dir) !void {
	var dataDir = cwd.openDir(DATA_DIR, .{.iterate = true}) catch |err| switch (err) {
		error.FileNotFound => {
			try stdout.print("Could not find data! Reinstalling...\n", .{});
			try stdout.flush();
			try runInstaller(allocator, .FIX);
			unreachable;
		},
		error.NotDir => {
			try stderr.print("Data is not a directory! Reinstalling...\n", .{});
			try runInstaller(allocator, .FIX);
			unreachable;
		},
		else => {
			try stderr.print("Could not open data directory: {}\n", .{err});
			std.process.exit(BAD_EXIT);
		}
	};
	defer dataDir.close();

	var dataDirWalker = try dataDir.walk(allocator);
	defer dataDirWalker.deinit();

	const numDirs = DATA_DIRS.len;
	var found: [numDirs]bool = undefined;
	for (0..numDirs) |i| {
		found[i] = false;
	}

	while (try dataDirWalker.next()) |entry| {
		if (entry.kind != .directory) continue;
		for (0..numDirs) |i| {
			const dir = DATA_DIRS[i];
			if (std.mem.eql(u8, entry.path, dir)) {
				std.debug.print("Found data directory: {s} from {s}\n", .{dir, entry.path});
				found[i] = true;
				break;
			}
		}
	}

	if (!found[2]) {
		// Not finding saves/ is not an error, just make it
		const savesFullPath = try std.fs.path.join(allocator, &[_][]const u8{DATA_DIR, DATA_DIRS[2]});
		cwd.makeDir(savesFullPath) catch |err| {
			try stderr.print("Could not create saves directory: {}\n", .{err});
			std.process.exit(BAD_EXIT);
		};
		found[2] = true;
	}

	var missing = false;
	for (0..numDirs) |i| {
		const dir = DATA_DIRS[i];
		if (!found[i]) {
			if (!missing) try stderr.print("Missing required data directories:\n", .{});
			missing = true;
			try stderr.print(" - {s}\n", .{dir});
		}
	}
	if (missing) {
		try stderr.flush();
		runInstaller(allocator, .FIX) catch { std.process.exit(BAD_EXIT); };
		unreachable;
	}
}

pub fn main() !void {
	var gpa = std.heap.GeneralPurposeAllocator(.{}).init;
	defer _ = gpa.deinit();
	const allocator = gpa.allocator();

	switch (builtin.os.tag) {
		.windows => {
			w = std.fs.File.stdout().writer(&outBuffer);
			e = std.fs.File.stderr().writer(&errBuffer);
		},
		else => {}
	}

	// check existence of version file
	const cwd = std.fs.cwd();
	const vf = cwd.openFile("version", .{ .mode = .read_only }) catch {
		// std.debug.print("Could not find version file! Fixing files...\n", .{});
		try stderr.print("Could not find version file! Fixing files...\n", .{});
		try stderr.flush();
		runInstaller(allocator, Opt.FIX) catch { std.process.exit(BAD_EXIT); };
		return;
	};
	vf.close();

	ssleep(1000);

	const latest = try checkLatest(allocator);
	if (!latest) {
		var rBuffer: [1024]u8 = undefined;
		var r: std.fs.File.Reader = std.fs.File.stdin().reader(&rBuffer);
		const stdin = &r.interface;

		try stdout.print("New version found! Do you want to update? [y|n] \n", .{});
		try stdout.flush();
		var res = stdin.takeByte() catch |err| {
			try stderr.print("Failed reading response: {}\n", .{err});
			return;
		};

		if (res >= 0x41 and res <= 0x5A) res += 32;
		if (res == 'y') {
			runInstaller(allocator, Opt.UPDATE) catch { std.process.exit(BAD_EXIT); };
			return;
		} else {
			try stdout.print("Will not update. It is highly recommended to update for latest features and fixes!\n", .{});
		}
	}

	try stdout.print("Verifying files and data...\n", .{});
	ssleep(1000);

	const gameFile = cwd.openFile(GAME, .{ .mode = .read_only }) catch {
		try stdout.print("Could not find game! Fixing files...\n", .{});
		try stdout.flush();
		runInstaller(allocator, Opt.FIX) catch { std.process.exit(BAD_EXIT); };
		return;
	};
	gameFile.close();

	try verifyDataDir(allocator, cwd);

	try stdout.print("Verification done. Game starting...\n", .{});
	try stdout.flush();
	ssleep(1000);

	const launcherFlag = "-l";

	const cwdStr = try std.fs.selfExeDirPathAlloc(allocator);
	defer allocator.free(cwdStr);

	const gameExe = try std.fs.path.join(allocator, &[_][]const u8{cwdStr, GAME});
	defer allocator.free(gameExe);
	const gameArgs = &[_][]const u8{gameExe, launcherFlag};

	var proc = std.process.Child.init(gameArgs, allocator);
	proc.cwd = cwdStr;
	// if (proc.cwd) |_cwd| {
	// std.debug.print("Setting working directory to {s}\n", .{_cwd});
	// } else {
	// std.debug.print("No working directory set for child process\n", .{});
	// }
	const exitCode = proc.spawnAndWait() catch |err| {
		try stdout.print("Could not start game: {}\n", .{err});
		try stdout.flush();
		std.process.exit(BAD_EXIT);
	};

	if (exitCode.Exited != 0) {
		try stdout.print("Game exited with code {}\n", .{exitCode});
		try stdout.flush();
		std.process.exit(BAD_EXIT);
	}
}

File: launcher.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include <errno.h>
#include <dirent.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <curl/curl.h>



#define BAD_EXIT 0xff // -1

#define INSTALLER "clisw-installer"
#define LAUNCHER "clisw-launcher"
#define GAME "clisw"

#define INSTALLER_DIR ".."
#define DATA_DIR "data"
#define MISC_DIR "data/misc"
#define STORY_DIR "data/story"
#define MAPS_DIR "data/maps"
#define SAVES_DIR "data/saves"


#define UNREACHABLE() do { printf("SHOULD NOT REACH THIS!\n"); exit(-1); } while (0);


enum OPTS {
  FIX,
  UPDATE,
  NONE
};

/**
 * Wrapper around sleep function for Windows and Linux.
 * @param ms Amount of milliseconds to sleep for
 */
static void ssleep(int ms) {
  usleep(ms * 1000);
}



size_t writeCallback(void* contents, size_t size, size_t nmemb, void* userp) {
  size_t realsize = size * nmemb;
  strncat((char*)userp, (char*)contents, realsize);
  return realsize;
}

bool checkLatest() {
  CURL* curl;
  CURLcode res;
  char remoteVersion[256] = {0};
  
  curl_global_init(CURL_GLOBAL_DEFAULT);
  curl = curl_easy_init();

  if (curl) {
    struct curl_slist* headers = NULL;
    headers = curl_slist_append(headers, "Accept: application/vnd.github.raw+json");
    headers = curl_slist_append(headers, "User-Agent: VersionChecker/1.0");

    curl_easy_setopt(curl, CURLOPT_URL, "https://api.github.com/repos/jgarcdev/TextAdventure-CLISoulWorker/contents/version");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);  // Follow redirects
    curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 5L);       // Max 5 redirects
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, remoteVersion);
    
    res = curl_easy_perform(curl);
    
    if (res != CURLE_OK) {
      fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
      curl_easy_cleanup(curl);
      curl_slist_free_all(headers);
      curl_global_cleanup();
      return -1;
    }

    curl_easy_cleanup(curl);
    curl_slist_free_all(headers);
  }
  curl_global_cleanup();

  // Read local version
  FILE* fp = fopen("version", "r");
  if (!fp) return -1;

  char localVersion[256] = {0};
  fgets(localVersion, sizeof(localVersion), fp);
  fclose(fp);

  // Trim newlines
  localVersion[strcspn(localVersion, "\r\n")] = 0;
  remoteVersion[strcspn(remoteVersion, "\r\n")] = 0;

  printf("Checking version...\n");
  printf("Current version is %s\n", localVersion);
  printf("Latest version is %s\n", remoteVersion);

  if (strcmp(remoteVersion, localVersion) == 0) return true;
  return false;
}

/**
 * Runs the installer based on the passed options.
 * @param opt The option for the installer
 */
static void runInstaller(enum OPTS opt) {
  printf("Changing to installer directory...\n");
  ssleep(1000);

  int ret = chdir(INSTALLER_DIR);

  if (ret == -1) {
    perror("ERROR");
    printf("Cannot change into installer directory. Exiting...\n");
    exit(-1);
  }

  char* args[3] = {INSTALLER, NULL, NULL};

  if (opt == FIX) args[1] = "-f";
  else args[1] = "-u";

  int execRet = execv(INSTALLER, args);

  if (execRet == -1) {
    perror("ERROR");
    printf("Could not execute installer!\n");
    exit(BAD_EXIT);
  }
}

int main(int argc, char const* argv[]) {

  // check existance of version file
  FILE* versionFile = fopen("version", "r");
  if (!versionFile) {
    printf("Could not find version file! Fixing files...\n");

    runInstaller(FIX);
    UNREACHABLE()
  }
  fclose(versionFile);

  ssleep(1000);

  bool latest = checkLatest();
  if (!latest) {
    printf("New version found! Do you want to update? [y|n] \n");
    char response = tolower(getchar());

    if (response == 'y') {
      runInstaller(UPDATE);
      UNREACHABLE()
    } else printf("Will not update. It is highly recommended to update for latest features and fixes!\n");
  }

  // will not update, now verify file integrity
  // for future, do it in a better way
  printf("Verifying files and data...\n");
  ssleep(1000);

  FILE* game = fopen(GAME, "r");
  if (!game) {
    printf("Could not find game! Fixing file...\n");
    runInstaller(FIX);
    UNREACHABLE()
  }
  fclose(game);


  DIR* data = opendir(DATA_DIR);
  if (!data) {
    printf("Could not find data! Reinstalling...\n");
    runInstaller(UPDATE);
    UNREACHABLE()
  }
  closedir(data);

  DIR* maps = opendir(MAPS_DIR);
  if (!maps) {
    printf("Could not find map data! Fixing files...\n");
    runInstaller(FIX);
    UNREACHABLE()
  }
  closedir(maps);

  DIR* misc = opendir(MISC_DIR);
  if (!misc) {
    printf("Could not find misc data! Fixing files...\n");
    runInstaller(FIX);
    UNREACHABLE()
  }
  closedir(misc);

  DIR* story = opendir(STORY_DIR);
  if (!story) {
    printf("Could not find story data! Fixing files...\n");
    runInstaller(FIX);
    UNREACHABLE()
  }
  closedir(story);

  // story = opendir("data/story/best_showtime");
  // if (!story) {
  //   printf("Could not find story data! Fixing files...\n");
  //   runInstaller(FIX);
  //   UNREACHABLE()
  // }
  // closedir(story);

  story = opendir("data/story/control_zone");
  if (!story) {
    printf("Could not find story data! Fixing files...\n");
    runInstaller(FIX);
    UNREACHABLE()
  }
  closedir(story);

  story = opendir("data/story/r_square");
  if (!story) {
    printf("Could not find story data! Fixing files...\n");
    runInstaller(FIX);
    UNREACHABLE()
  }
  closedir(story);

  // story = opendir("data/story/tower_of_greed");
  // if (!story) {
  //   printf("Could not find story data! Fixing files...\n");
  //   runInstaller(FIX);
  //   UNREACHABLE()
  // }
  // closedir(story);

  int ret;

  DIR* saves = opendir(SAVES_DIR);
  if (!saves) {
    printf("Could not find save data directory! Creating saves directory...\n");

    ret = mkdir(SAVES_DIR, S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);

    if (ret == -1) {
      if (errno != EEXIST) {
        perror("ERROR");
        printf("Cannot create saves directory. Exiting...\n");
        exit(-1);
      }
    }
  }
  closedir(saves);

  printf("Verification done. Game starting...\n");
  ssleep(1000);

  const char* launcherFlag = "-l";
  int execRet = execl(GAME, GAME, launcherFlag, NULL);
  if (execRet == -1) {
    printf("COULD NOT EXECUTE CLISW, ABORTING\n");
    exit(-1);
  }

  return 0;
}